### PR TITLE
feat: FastEmbed embedding provider

### PR DESCRIPTION
**Pull request by Anush008**
## Description:
This PR intends to add [Qdrant/FastEmbed](https://qdrant.github.io/fastembed/) as a local embeddings provider, associated tests and documentation.

**Documentation preview:** https://langchain-git-fork-anush008-master-langchain.vercel.app/docs/integrations/text_embedding/fastembed

### CODE REVIEW

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    output = await embedding.aembed_documents(documents)
+    assert len(output) == 2
+    assert len(output[0]) == 384
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "model_name", ["sentence-transformers/all-MiniLM-L6-v2", "BAAI/bge-small-en-v1.5"]
+)
+@pytest.mark.parametrize("max_length", [50, 512])
```
Comment (KShivendu) [inline libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:68]:
```suggestion
@pytest.mark.parametrize(
    "model_name,max_length",
    [
        ("sentence-transformers/all-MiniLM-L6-v2", 50),
        ("BAAI/bge-small-en-v1.5", 512)
        # Add other combinations as needed
    ]
)
```

You can cover them in one go.

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    assert len(output) == 384
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "model_name", ["sentence-transformers/all-MiniLM-L6-v2", "BAAI/bge-small-en-v1.5"]
+)
+@pytest.mark.parametrize("max_length", [50, 512])
+@pytest.mark.parametrize("doc_embed_type", ["default", "passage"])
+@pytest.mark.parametrize("threads", [0, 10])
```
Comment (KShivendu) [inline libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:47]:
```suggestion
@pytest.mark.parametrize(
    "model_name, max_length, doc_embed_type, threads",
    [
        ("sentence-transformers/all-MiniLM-L6-v2", 50, "default", 0),
        ("sentence-transformers/all-MiniLM-L6-v2", 50, "passage", 10),
        ("BAAI/bge-small-en-v1.5", 512, "default", 0),
        ("BAAI/bge-small-en-v1.5", 512, "passage", 10),
        # Add other specific combinations as needed
    ]
)
```

Try this.

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+from langchain.embeddings.fastembed import FastEmbedEmbeddings
+
+
+@pytest.mark.parametrize(
+    "model_name", ["sentence-transformers/all-MiniLM-L6-v2", "BAAI/bge-small-en-v1.5"]
+)
+@pytest.mark.parametrize("max_length", [50, 512])
+@pytest.mark.parametrize("doc_embed_type", ["default", "passage"])
+@pytest.mark.parametrize("threads", [0, 10])
```
Comment (KShivendu) [inline libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:12]: You can do this in one go. See next set of comments.

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py (hunk 2 above)
Comment (Anush008) [reply libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:47]: Wouldn't this be more typing? And specifying them individually will cover all the permutations without having to specify all of them.

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py (hunk 1 above)
Comment (Anush008) [reply libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:68]: In the current state, all the permutations will be checked without having to specify them explicitly.

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py (hunk 2 above)
Comment (KShivendu) [reply libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:47]: Ahh  I see. Maybe try `from itertools import product` to create combinations. Or keep it your way.

File: libs/langchain/tests/integration_tests/embeddings/test_fastembed.py (hunk 1 above)
Comment (KShivendu) [reply libs/langchain/tests/integration_tests/embeddings/test_fastembed.py:68]: https://github.com/langchain-ai/langchain/pull/13109#discussion_r1387990241

File: libs/langchain/langchain/embeddings/fastembed.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+import numpy as np
+
+from langchain.pydantic_v1 import BaseModel, Extra, root_validator
+from langchain.schema.embeddings import Embeddings
+
+
+class FastEmbedEmbeddings(BaseModel, Embeddings):
+    """Qdrant FastEmbedding models.
+
```
Comment (eyurtsev) [inline libs/langchain/langchain/embeddings/fastembed.py:11]: Could we add links to documentations -- it's easier for maintainers to provide support for packages they're not familiar with

```suggestion
    FastEmbed is a lightweight, fast, Python library built for embedding generation.
    See more documentation at: 
    * https://github.com/qdrant/fastembed/
    * https://qdrant.github.io/fastembed/

    To use this class, you must install the `fastembed` Python package.

    `pip install fastembed`
```

**Review by eyurtsev — COMMENTED** [review]
Looks good, can merge after resolving tests

**Comment by Anush008** [comment]
@eyurtsev, should be good now.

**Comment by eyurtsev** [comment]
Can merge if passes tests

**Pull request merged**